Include untracked time in composite trips + refactor confirmed object creation

JGreenlee opened the pull request:
This enables https://github.com/e-mission/e-mission-phone/pull/935.

In the process, it also:
- Significantly refactors confirmed object creation. Instead of handling confirmed places and confirmed trips separately, we will use the handy `Timeline` from emission>storage>decorators>timeline.
- - For this to work, `Timeline` had to be slightly modified / added to. Since we want a timeline of expected trips and cleaned places, and the cleaned places only have references to cleaned trips, we need to identify the expected trips by the cleaned_trip they were created from. This is accomplished by a new optional argument.
- The refactor fixes the issue of confirmed places sometimes not matching to the appropriate trip, which would happen in the case of untracked time.

shankari committed:
Tweak the threshold for detecting untracked time to account for background restrictions

After all the background restrictions, particularly on android, we can have a
pretty large gap in trip start detection. The original 1k meter threshold for
detecting untracked time was a rough guess, and is now too small.

On looking at staging data, we empirically find that:
- we see untracked time more often on android (https://github.com/e-mission/e-mission-docs/issues/870#issuecomment-1484193103)
- the IQR for iOS is fairly close to 1k, but the IQR for android is now closer to 2.5k
https://github.com/e-mission/e-mission-docs/issues/870#issuecomment-1484199825
- ~ 50% of the untracked time on android does not have any associated reboots
https://github.com/e-mission/e-mission-docs/issues/870#issuecomment-1484214939

So we:
- split the implementation between android and iOS
- bump up the threshold on iOS to 1.2k and on android to 2.5k
https://github.com/e-mission/e-mission-docs/issues/870#issuecomment-1484199825

Testing done:
- All real world tests pass

JGreenlee committed:
allow timeline to support IDing trips by a key

JGreenlee committed:
add real_examples that demonstrate untracked time

JGreenlee committed:
refactor confirmed object creation to use Timeline

JGreenlee committed:
include untracked time in composite trip creation

JGreenlee committed:
don't return early from create_composite_objects

if there are no new composite trips to be created, we want to mark the stage as done with value None

shankari commented on emission/analysis/plotting/composite_trip_creation.py:
```diff
@@ -1,5 +1,5 @@
 import emission.core.wrapper.entry as ecwe
-from emission.analysis.userinput.matcher import create_confirmed_place_entry, get_confirmed_place_for_confirmed_trip
+import emission.analysis.userinput.matcher as eaum
```
glad to see that you made this consistent with the rest of the codebase!

shankari commented on emission/analysis/plotting/composite_trip_creation.py:
```diff
@@ -10,26 +10,28 @@
-    composite_trip_dict["data"]["locations"] = get_locations_for_confirmed_trip(ct)
-    # later we will want to put section & modes in composite_trip as well
+    if not isUntrackedTime:
+        composite_trip_dict["data"]["confirmed_place"] = eaum.get_confirmed_place_for_confirmed_trip(ct)
```
I know that cleaned untracked time will have a cleaned end place. Why do you not have that for composite untracked time? Please add high-level comment in source code as well.

JGreenlee replied:
Yes, but that place will have a duration of 0, and be ineligible for additions or userinputs, so I did not think we wanted to display it.

Do you think it needs to be shown?

shankari replied:
I don't think it needs to be shown, but I was wondering if it would break other downstream assumptions about whether or not trips have places. If it will not, I am fine with adding a comment here, and adding test cases for this behavior so that we can be notified if/when it changes.

JGreenlee replied:
Added comment above.
The Jack Mar 12 test will check the overall output of composite trips, so if something does change, that test can tell us.

shankari commented on emission/analysis/userinput/matcher.py:
```diff
@@ -76,83 +77,60 @@ def handle_multi_non_deleted_match(confirmed_obj, ui):
+                    logging.debug("Found trip that came before place, inserting confirmed trip with confirmed place id")
+                    cTrip['data']['confirmed_place'] = cp_id
+                    ts.insert(cTrip)
+                elif untracked:
+                    logging.debug("Found place that follows untracked time, confirmed trip will not be created")
```
I don't understand this. `untracked_time` is intended to be `trip_like` with a start and end timestamp. So why do you log this as a place?
So the timeline around untracked time should be 

```
... -> cleaned place -> expected trip -> cleaned place -> untracked time -> cleaned place -> expected trip -> ...
```

If I understand this logic correctly, you will end up with 

```
... -> confirmed place -> confirmed trip -> confirmed place ->  -> confirmed place -> confirmed trip -> ...

```

Is that what you expect? I would like to see some unit tests around this code and the new real_examples. You can use `emission/tests/analysisTests/intakeTests/TestPipelineRealData.py` for an example of how to load data and run the pipeline.

You can check the results manually or automatically against a saved result.

JGreenlee replied:
I will reword that log statement - I see how it is confusing.

>Is that what you expect?

Yes, because I thought we did not want to create a confirmed object for untracked time. When composite trips are created, we just use the cleaned_untracked object

I will make unit tests to verify this is happening

shankari commented on emission/analysis/userinput/matcher.py:
```diff
@@ -76,83 +77,60 @@ def handle_multi_non_deleted_match(confirmed_obj, ui):
-        lastTripProcessed = confirmed_trip_entry
-
-    return lastTripProcessed
+def create_confirmed_entry(ts, tce, confirmed_key, input_key_list):
+    # Copy the entry and fill in the new values
+    confirmed_trip_dict = copy.copy(tce)
+    del confirmed_trip_dict["_id"]
```
nice refactoring!

shankari reviewed (changes requested):
Don't see any high level issues except the one around the places before and after the untracked time. I do think it would be good to add basic unit tests documenting what you expect to happen, and ensuring that it does happen, at least on the examples that we are committing here.

JGreenlee committed:
Merge branch 'add_trip_place_additions' into untracked-composite-trip

JGreenlee committed:
include origin_key in composite trip

Since composite trips may now be created from either 'confirmed_trip` or `cleaned_untracked` objects, we will insert this `origin_key`.
This allows the phone, upon receipt of a composite trip, to explicitly disambiguate the datatype that the composite trip was created from.

JGreenlee committed:
Merge branch 'add_trip_place_additions' into untracked-composite-trip

JGreenlee committed:
declare composite trip origin_key before key

JGreenlee committed:
if using trip_id_key in Timeline, fallback to _id

When getting a timeline of cleaned_places and expected_trips, we use trip_id_key to identify the expected_trips by the cleaned_trip they came from.
But if there are cleaned_untracked entries in this mix, we don't need to do this on those. So, if the trip_id_key is not present, we will just use the entry's _id via getId().

JGreenlee committed:
include empty locations list in untracked composite

JGreenlee committed:
reword comments in create_confirmed_objects

JGreenlee committed:
create testJackUntrackedTimeMar12

JGreenlee committed:
add more unit tests for untracked time

JGreenlee commented:
@shankari I wrote tests using Jack March 12, Shankari Mar 21, Shankari Jan 15, and Shankari Jul 20. There is trouble on the last.

After cherry-picking df9d9f0844eedcf7405d88afe9da1b02ee365986:
- Jack March 12 has untracked time. (expected)
- Shankari Mar 21 no longer has untracked time. (expected)
- Shankari Jan 15 has a reboot. This yields an instance of untracked time. (expected)
- Shankari Jul 20 has a large gap, but this is not resulting in untracked time. (not expected)

Thus, the first three tests will pass and the fourth test will fail.
Is this something we should worry about?

JGreenlee committed:
add comment explaining places after untracked time

shankari commented on emission/analysis/plotting/composite_trip_creation.py:
```diff
@@ -26,8 +26,8 @@ def create_composite_trip(ts, ct):
     del composite_trip_dict["_id"]
-    composite_trip_dict["metadata"]["key"] = "analysis/composite_trip"
     composite_trip_dict["metadata"]["origin_key"] = ct["metadata"]["key"]
+    composite_trip_dict["metadata"]["key"] = "analysis/composite_trip"
```
why is this important?

JGreenlee replied:
`origin_key` has to be assigned first because it should be the old value of `key`.
I had mistakenly assigned `key` first, in which case both would just be set to the new value `'analysis/composite_trip'`.

shankari reviewed (changes requested):
nit: please explain the change below for the record.
I am fine with merging this as such **except** that 

> Shankari Jul 20 has a large gap, but this is not resulting in untracked time. (not expected)

That is really weird, especially since it did generate untracked time earlier. 

```
2022-07-20T17:03:51-07:00 | 2022-07-20T18:43:39.669000-07:00 | 3483.113972
```

While waiting for @JGreenlee to address the comments in https://github.com/e-mission/e-mission-phone/pull/935, I will investigate why. If I can't figure it out, we can comment out/remove test 4 for now to address 

> Thus, the first three tests will pass and the fourth test will fail.

shankari commented:
We do create a raw_untracked object

```
2023-04-03 21:39:02,143:DEBUG:140704519095936:while determining new_start_place, transition_distance = 3483.1139715358286
2023-04-03 21:39:02,143:DEBUG:140704519095936:while determining new_start_place, time_delta = 5988.668999910355, transition_speed = 0.5816173796861988
2023-04-03 21:39:02,143:DEBUG:140704519095936:transition_speed 0.5816173796861988 <= 0.8333333333333334, 'stopped', returning True
```

```
2023-04-03 21:39:02
147:DEBUG:140704519095936:Inserting entry Entry({'_id': ObjectId('642ba9e606cb2e88da76540e')
 'user_id': UUID('21242744-799d-4967-a312-6c9c0f0c7776')
 'metadata': Metadata({'key': 'segmentation/raw_untracked'
 'platform': 'server'
 'write_ts': 1680583142.143456
 'time_zone': 'America/Los_Angeles'
 'write_fmt_time': '2023-04-03T21:39:02.143456-07:00'})
 'data': Untrackedtime({'source': 'DwellSegmentationTimeFilter'
 'start_ts': 1658361836
 'start_fmt_time': '2022-07-20T17:03:51-07:00'
 'start_place': ObjectId('642ba9e606cb2e88da76540b')
 'end_ts': 1658367819.669
 'end_local_dt': LocalDate({'year': 2022
 'month': 7
 'day': 20
 'hour': 18
 'minute': 43
 'second': 39
 'weekday': 2
 'timezone': 'America/Los_Angeles'})
 'end_fmt_time': '2022-07-20T18:43:39.669000-07:00'
 'end_place': ObjectId('642ba9e606cb2e88da76540f')
 'duration': 5983.668999910355
 'distance': 3483.1139715358286})}) into timeseries

2023-04-03 21:39:02,153:DEBUG:140704519095936:Created untracked period 642ba9e606cb2e88da76540e from 1658361836 to 1658367819.669
```

But when we are creating composite trips, we don't see a trip-like object corresponding to this

```
2023-04-03 21:39:04,275:DEBUG:140704519095936:Considering trip 642ba9e806cb2e88da76550d: 2022-07-20T12:03:26-07:00 -> 2022-07-20T12:13:12-07:00
2023-04-03 21:39:04,275:DEBUG:140704519095936:Considering trip 642ba9e806cb2e88da76550e: 2022-07-20T12:48:16.482150-07:00 -> 2022-07-20T14:04:24-07:00
2023-04-03 21:39:04,275:DEBUG:140704519095936:Considering trip 642ba9e806cb2e88da76550f: 2022-07-20T16:50:28.404757-07:00 -> 2022-07-20T17:03:51-07:00
```

Note also that the original untracked time was from

```
2022-07-20T17:03:51-07:00 | 2022-07-20T18:43:39.669000-07:00 | 3483.113972
```

Is this because the untracked time is at the end of the chain?

Let's look at the `CLEAN_AND_RESAMPLE` stage and see what happened to id `642ba9e606cb2e88da76540e`?

So we find the untracked time and the subsequent trip-like object

```
2023-04-03 21:39:02,660:DEBUG:140704519095936:Considering trip 642ba9e606cb2e88da76540a: 2022-07-20T16:50:41.970000-07:00 -> 2022-07-20T17:03:51-07:00
2023-04-03 21:39:02,660:DEBUG:140704519095936:Considering trip 642ba9e606cb2e88da76540e: 2022-07-20T17:03:51-07:00 -> 2022-07-20T18:43:39.669000-07:00
2023-04-03 21:39:02,660:DEBUG:140704519095936:Considering trip 642ba9e606cb2e88da76540c: 2022-07-20T18:43:39.669000-07:00 -> 2022-07-20T18:46:12-07:00
```

And we map it to a cleaned trip-like object

```
2023-04-03 21:39:03,661:DEBUG:140704519095936:Found untracked entry 642ba9e606cb2e88da76540e (2022-07-20T17:03:51-07:00 -> 2022-07-20T18:43:39.669000-07:00), skipping processing
2023-04-03 21:39:03,661:DEBUG:140704519095936:For raw trip 642ba9e606cb2e88da76540e, found filtered trip 642ba9e706cb2e88da7654f0
```

And then we detect that it is a squished trip, so we remove it

```
2023-04-03 21:39:03,842:DEBUG:140704519095936:Found squished trip 642ba9e606cb2e88da76540e, linking raw start place 642ba9e606cb2e88da76540b to existing cleaned place 642ba9e706cb2e88da7654f4
2023-04-03 21:39:03,842:DEBUG:140704519095936:Found squished trip 642ba9e606cb2e88da76540c, linking raw start place 642ba9e606cb2e88da76540f to existing cleaned place 642ba9e706cb2e88da7654f4
```

shankari commented:
I went back and checked this trip on staging.
There is a `raw_untracked` trip-like object for it, but no `cleaned_untracked`
I just queried `raw_untracked` since I was tweaking the trip segmentation stage
So this is not a regression/change

We might want to change this to reflect the cleaned untracked time.
However, similar to https://github.com/e-mission/e-mission-phone/pull/935, I don't want to change unrelated code, that is no obviously broken, at this time. This is especially true for fragile code like the trip squishing.

So we should change the test to reflect that test case #4 is supposed to have raw untracked time but no cleaned untracked time, and we can investigate further later.

shankari commented on emission/tests/analysisTests/intakeTests/TestPipelineRealData.py:
```diff
@@ -697,6 +698,71 @@ def testJan16SpeedAssert(self):
+        ts = esta.TimeSeries.get_time_series(self.testUUID)
+        composite_trips = list(ts.find_entries(["analysis/composite_trip"], None))
+        countUntrackedTime = 0
+        for ct in composite_trips:
+            if ct['metadata']['origin_key'] == 'analysis/cleaned_untracked':
+                countUntrackedTime += 1
+        self.assertEqual(countUntrackedTime, 1)
```
```suggestion
        for ct in composite_trips:
            if ct['metadata']['origin_key'] == 'analysis/cleaned_untracked':
                cleanUntrackedTime += 1
            elif ct['metadata']['origin_key'] == 'segmentation/raw_untracked':
                rawUntrackedTime += 1
        self.assertEqual(rawUntrackedTime, 1)
        self.assertEqual(cleanUntrackedTime, 0)
```

shankari commented:
@JGreenlee I am happy to merge this after the test case suggestion is committed.

JGreenlee committed:
update expected test result: shankariUntrackedTimeJul20

This test day will result in one `raw_untracked`, but no `cleaned_untracked`
https://github.com/shankari/e-mission-server/pull/7#issuecomment-1495356024

Co-authored-by: shankari <shankari@eecs.berkeley.edu>

JGreenlee commented:
Good to go.
